### PR TITLE
ext.tasks.loop inheritance support

### PR DESCRIPTION
**Pull request by Enegg**
## Summary
Supercedes and closes #259

Fixes an issue with `Loop` descriptor returning `Loop` instance in child classes; `loop` decorator supports `Loop` subclasses
## Checklist

- [ ] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [ ] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by onerandomusername** [comment]
@Enegg will you be resolving the issues with this pr?

File: changelog/655.feature.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Add support for subclassing ``.ext.tasks.Loop`` and using subclasses in :func:`.ext.tasks.loop` decorator
```
Comment (onerandomusername) [inline changelog/655.feature.rst:1]:
```suggestion
|tasks| Add support for subclassing :class:`.ext.tasks.Loop` and using subclasses in :func:`.ext.tasks.loop` decorator
```

File: disnake/ext/tasks/__init__.py [hunk 2]
```diff
@@ -73,31 +100,35 @@ def cancel(self) -> None:
...
-        reconnect: bool,
-        loop: asyncio.AbstractEventLoop,
+        coro: Coro[CoroP],
+        seconds: float = 0,
+        minutes: float = 0,
+        hours: float = 0,
+        time: Union[datetime.time, Sequence[datetime.time]] = MISSING,
+        count: Optional[int] = None,
+        reconnect: bool = True,
+        loop: asyncio.AbstractEventLoop = MISSING,
```
Comment (onerandomusername) [inline disnake/ext/tasks/__init__.py:118]: might be worth making these keyword only arguments in this same pr, since they're currently being reordered which is already breaking

File: disnake/ext/tasks/__init__.py [hunk 3]
```diff
@@ -73,31 +100,35 @@ def cancel(self) -> None:
...
     ) -> None:
-        self.coro: LF = coro
+        """
+        .. Note:
+            If you're going to overwrite __init__ arguments, make sure to redefine .clone too
+        """
```
Comment (onerandomusername) [inline disnake/ext/tasks/__init__.py:123]: i don't think this appears on the documentation, the docstring for `Loop` needs to be updated with this instead.

File: changelog/655.feature.rst (hunk 1 above)
Comment (onerandomusername) [reply changelog/655.feature.rst:1]: Additionally, if the arguments are made into keyword arguments, that should be added as `655.breaking.rst`.

File: disnake/ext/tasks/__init__.py [hunk 4]
```diff
@@ -695,21 +730,55 @@ def change_interval(
...
+    **kwargs: Any,
+) -> Callable[[Coro[CoroP]], L_co]:
     """A decorator that schedules a task in the background for you with
     optional reconnect logic. The decorator returns a :class:`Loop`.
 
     Parameters
     ----------
+    cls: Type[:class:`Loop`]
+        The loop subclass to create an instance of. If provided, the following parameters
+        described below do no apply. Instead, this decorator will accept the same keywords
```
Comment (onerandomusername) [inline disnake/ext/tasks/__init__.py:777]:
```suggestion
        described below do not apply. Instead, this decorator will accept the same keywords
```

**Review by onerandomusername — CHANGES_REQUESTED** [review]
thanks so far for what you've shared, these are mostly documentation notes or an easy fix.

File: disnake/ext/tasks/__init__.py (hunk 3 above)
Comment (Enegg) [reply disnake/ext/tasks/__init__.py:123]: The intent was for this *not to* be visible on the Loop itself, and instead on code introspection; I don't think any class documents any info like that

Comment (onerandomusername) [reply disnake/ext/tasks/__init__.py:123]: makes sense, but in that case the rst formatting is unused.

Comment (Enegg) [reply disnake/ext/tasks/__init__.py:124]: So you suggest to remove the formatting? It will look kind of out of place when a "short description of the function" is actually a single note

Comment (onerandomusername) [reply disnake/ext/tasks/__init__.py:124]: fair. if we're committing to rst we might as well make it valid.
```suggestion
        """
        .. note:
            If you overwrite ``__init__`` arguments, make sure to redefine .clone too.
        """
```

**Review by onerandomusername — APPROVED** [review]
lgtm

File: disnake/ext/tasks/__init__.py [hunk 5]
```diff
@@ -695,21 +731,55 @@ def change_interval(
                 self._handle.recalculate(self._next_iteration)
 
 
+T_co = TypeVar("T_co", covariant=True)
+L_co = TypeVar("L_co", bound=Loop, covariant=True)
+
+
+class Object(Protocol[T_co, P]):
+    def __new__(cls) -> T_co:
```
Comment (shiftinv) [inline disnake/ext/tasks/__init__.py:739]:
```suggestion
L_co = TypeVar("L_co", bound=Loop, covariant=True)


class Object(Protocol[L_co, P]):
    def __new__(cls) -> L_co:
```

File: disnake/ext/tasks/__init__.py [hunk 6]
```diff
@@ -745,20 +815,21 @@ def loop(
...
+        cls = origin
+
+    if not isinstance(cls, type) or not issubclass(cls, Loop):
+        raise TypeError(f"cls argument must be a subclass of Loop, got {cls!r}")
+
+    def decorator(func: Coro[CoroP]) -> L_co:
+        if not asyncio.iscoroutinefunction(func):
+            raise TypeError("decorated function must be a coroutine")
+
+        return cast(Type[L_co], cls)(func, **kwargs)
```
Comment (shiftinv) [inline disnake/ext/tasks/__init__.py:833]:
```suggestion
        return cast("Type[L_co]", cls)(func, **kwargs)
```

File: disnake/ext/tasks/__init__.py [hunk 7]
```diff
@@ -31,20 +31,47 @@
...
 import disnake
 from disnake.backoff import ExponentialBackoff
 from disnake.utils import MISSING, utcnow
 
+if TYPE_CHECKING:
+    from typing_extensions import Concatenate, ParamSpec, Self
+
+    P = ParamSpec("P")
+    CoroP = ParamSpec("CoroP")
+    Coro = Callable[CoroP, Coroutine[Any, Any, Any]]
```
Comment (shiftinv) [inline disnake/ext/tasks/__init__.py:62]: This should strip the self type from `CoroP`, otherwise type-checkers will expect you to use `self.background_task(self, x, y)` in classes (which doesn't work) instead of `self.background_task(x, y)`.

This also came up [here](https://github.com/DisnakeDev/disnake/pull/675#discussion_r968657205) a few days ago, however that one's a bit more clear as it only needs to account for `self` being a `Cog` instance, while `Loop` could technically be used without cogs entirely.
I don't think there's a way of reliably stripping the self type from a paramspec regardless of the type, so maybe it's sufficient for now to only account for cogs as well.

tl;dr, something along the lines of this would be needed
```py
    InjectionCallback = Union[
        Callable[Concatenate[CogT, P], T_],
        Callable[P, T_],
    ]
```

**Review by shiftinv — CHANGES_REQUESTED** [review]
`Loop.start` and `Loop.restart` currently use `*args: Any, **kwargs: Any`, those should make use of the ParamSpec as well.

File: disnake/ext/tasks/__init__.py (hunk 5 above)
Comment (Enegg) [reply disnake/ext/tasks/__init__.py:739]: It's made to be generic for anything to comply with the same implementation I made for `ui.button` & `ui.select`.
Ideally the `Object` would be moved from both here and `ui.item` to some kind of disnake typeshed, but that was out of the scope of either PR

Comment (shiftinv) [reply disnake/ext/tasks/__init__.py:739]: I see, fair enough.

File: disnake/ext/tasks/__init__.py (hunk 7 above)
Comment (Enegg) [reply disnake/ext/tasks/__init__.py:62]: I'm looking at the examples, but I don't think it can be done in a manner that'd work with inheritance; the `__get__` would need to look something like
```py
@overload
def __get__(self, instance: None, owner: Any) -> Loop[CoroP]:
    ...
@overload
def __get__(self, instance: T, owner, Any) -> Loop[Concatenate[T, CoroP]]:
    ....
```
...which in both definitions hardcodes the return obj type as `Loop`
*There \*might\* be some sneaky way using Protocols (my beloved) to hint the return type as `SomeProto[Self, CoroP]` but I don't know how it is going to work here*

**Comment by Enegg** [comment]
Uh-oh, cannot make the `.start` and `.restart` use the ParamSpec without previously ressolving the stripping `self` param problem.

File: disnake/ext/tasks/__init__.py (hunk 7 above)
Comment (shiftinv) [reply disnake/ext/tasks/__init__.py:62]: Yeahh exactly, that'd work if `LoopT[CoroP]` was a thing or if one could rebind `Self`, which unfortunately isn't the case.
The [PEP](https://peps.python.org/pep-0673/) even explicitly states this:
> Note that we reject using Self with type arguments, such as Self[int].
> ...
> In such cases, we recommend using an explicit type for self

... which doesn't work, since we *don't* want an explicit type for self here :/

I've been trying to find an alternative solution, spent the last hour trying different things and looking through mypy/pyright bugtrackers and stackoverflow, but it seems this just isn't possible without HKT.

For now it's probably best to go back to `Loop[LF]`, and type all the `*args` and `**kwargs` as `Any`.

**Comment by onerandomusername** [comment]
> Uh-oh, cannot make the `.start` and `.restart` use the ParamSpec without previously ressolving the stripping `self` param problem.

@shiftinv will you be fixing this?

File: disnake/ext/tasks/__init__.py (hunk 7 above)
Comment (Enegg) [reply disnake/ext/tasks/__init__.py]: Done as suggested

Comment (onerandomusername) [reply disnake/ext/tasks/__init__.py]: https://github.com/DisnakeDev/disnake/pull/655/commits/2215240044a4fdd379a891b2e1681fa5525c7caf

**Review by shiftinv — APPROVED** [review]
